src/gui/transferlistsortmodel.cpp

```diff
@@ -86,179 +86,157 @@ void TransferListSortModel::disableTrackerFilter()
     Q_ASSERT(left.column() == right.column());
 
-    const auto invokeLessThanForColumn = [this, &left, &right](const int column) -> bool
-    {
-        return lessThan_impl(left.sibling(left.row(), column), right.sibling(right.row(), column));
-    };
-
-    const auto hashLessThan = [this, &left, &right]() -> bool
-    {
-        const TransferListModel *model = qobject_cast<TransferListModel *>(sourceModel());
-        const BitTorrent::InfoHash hashL = model->torrentHandle(left)->hash();
-        const BitTorrent::InfoHash hashR = model->torrentHandle(right)->hash();
-        return hashL < hashR;
-    };
-
     const int sortColumn = left.column();
     const QVariant leftValue = left.data(TransferListModel::UnderlyingDataRole);
     const QVariant rightValue = right.data(TransferListModel::UnderlyingDataRole);
 
     switch (sortColumn)
     {
     case TransferListModel::TR_CATEGORY:
-    case TransferListModel::TR_TAGS:
     case TransferListModel::TR_NAME:
-        if (!leftValue.isValid() || !rightValue.isValid() || (leftValue == rightValue))
-            return invokeLessThanForColumn(TransferListModel::TR_QUEUE_POSITION);
-        return (Utils::String::naturalCompare(leftValue.toString(), rightValue.toString(), Qt::CaseInsensitive) < 0);
+    case TransferListModel::TR_SAVE_PATH:
+    case TransferListModel::TR_TAGS:
+    case TransferListModel::TR_TRACKER:
+        {
+            const auto left = leftValue.toString();
+            const auto right = rightValue.toString();
+
+            if (left != right)
+                return Utils::String::naturalCompare(left, right, Qt::CaseInsensitive) < 0;
+        }
+        break;
+
+    case TransferListModel::TR_AMOUNT_DOWNLOADED:
+    case TransferListModel::TR_AMOUNT_DOWNLOADED_SESSION:
+    case TransferListModel::TR_AMOUNT_LEFT:
+    case TransferListModel::TR_AMOUNT_UPLOADED:
+    case TransferListModel::TR_AMOUNT_UPLOADED_SESSION:
+    case TransferListModel::TR_COMPLETED:
+    case TransferListModel::TR_ETA:
+    case TransferListModel::TR_LAST_ACTIVITY:
+    case TransferListModel::TR_SIZE:
+    case TransferListModel::TR_TIME_ELAPSED:
+    case TransferListModel::TR_TOTAL_SIZE:
+        {
+            const auto left = leftValue.toLongLong();
+            const auto right = rightValue.toLongLong();
+
+            if ((left < 0) && (right < 0))
+                break;
+            else if (right < 0)
+                return true;
+            else if (left < 0)
+                return false;
+            else if (left != right)
+                return left < right;
+        }
+        break;
+
+    case TransferListModel::TR_AVAILABILITY:
+    case TransferListModel::TR_PROGRESS:
+    case TransferListModel::TR_RATIO:
+    case TransferListModel::TR_RATIO_LIMIT:
+        {
+            const auto left = leftValue.toReal();
+            const auto right = rightValue.toReal();
+
+            if ((left < 0) && (right < 0))
+                break;
+            else if (right < 0)
```

Comment:
Negative values are interpreted as greater than any positive ones. Is it intended?

Reply:
>Negative values are interpreted as greater than any positive ones. Is it intended?

You read wrong. For the case here, left is placed *before* right.

Reply:
>Is it intended?

Yes, the negative values are considered to be invalid. So we place valid values before them.

Reply:
>Yes, the negative values are considered to be invalid. So we place valid values before them.

Then shouldn't they be placed after valid when reversed order is used?

Reply:
>Then shouldn't they be placed after valid when reversed order is used?

Yes, they would be reversed.
I find this question somewhat weird. The `lessThan()` don't care what order (ascending, descending) is in use, it only cares about comparison between 2 values and we always place valid values first and then comes the invalid ones.

Reply:
Either you didn't understand my question, or I didn't understand your answer...
If we take the following set of numbers, how are they supposed to be arranged in forward and reverse order (given that negative ones are treated as invalid):
1, 5, -1, 3, -1, 7, 2, -1
There are two ways:
1. Invalid are always in the same place, either in the beginning or end of list:
1, 2, 3, 5, 7, -1, -1, -1 (forward) and 7, 5, 3, 2, 1, -1, -1, -1 (reversed)
2. Place of invalid values depends from the sorting order:
1, 2, 3, 5, 7, -1, -1, -1 (forward) and -1, -1, -1, 7, 5, 3, 2, 1 (reversed)

Reply:
>Place of invalid values depends from the sorting order:
1, 2, 3, 5, 7, -1, -1, -1 (forward) and -1, -1, -1, 7, 5, 3, 2, 1 (reversed)

qbt needs this behavior. We have infinity symbols that are represented by negative values and this order is used throughout the years.